File: backend/app/games/sticks.py
"""Jeu des bâtonnets (Nim en version « le dernier perd »).

Règles : N bâtonnets au départ, chaque joueur en retire 1 à 3, celui qui prend
le dernier a perdu. Le jeu est résolu : laisser à l'adversaire un nombre de
bâtonnets congru à 1 modulo 4 garantit la victoire.
"""

from dataclasses import dataclass, replace

from app.games.base import IllegalMoveError, Player

MIN_STICKS = 5
MAX_STICKS = 50
DEFAULT_STICKS = 21
MAX_TAKE = 3


@dataclass(frozen=True)
class SticksState:
    sticks: int
    current: Player
    winner: Player | None = None

    @property
    def finished(self) -> bool:
        return self.winner is not None


def new_game(sticks: int = DEFAULT_STICKS, first: Player = Player.PLAYER) -> SticksState:
    if not MIN_STICKS <= sticks <= MAX_STICKS:
        raise ValueError(f"sticks must be between {MIN_STICKS} and {MAX_STICKS}, got {sticks}")
    return SticksState(sticks=sticks, current=first)


def legal_moves(state: SticksState) -> list[int]:
    if state.finished:
        return []
    return list(range(1, min(MAX_TAKE, state.sticks) + 1))


def apply(state: SticksState, take: int) -> SticksState:
    if state.finished:
        raise IllegalMoveError("game is finished")
    if take not in legal_moves(state):
        raise IllegalMoveError(f"cannot take {take} sticks with {state.sticks} remaining")
    remaining = state.sticks - take
    if remaining == 0:
        # Celui qui prend le dernier bâtonnet perd.
        return replace(state, sticks=0, winner=state.current.other)
    return replace(state, sticks=remaining, current=state.current.other)


def is_losing_position(sticks: int) -> bool:
    """Vrai si le joueur qui doit jouer avec `sticks` bâtonnets perd face à un jeu parfait."""
    return sticks % 4 == 1


def winning_move(sticks: int) -> int | None:
    """Coup qui laisse 4k+1 bâtonnets à l'adversaire, ou None en position perdante."""
    take = (sticks - 1) % 4
    return take if 1 <= take <= MAX_TAKE else None

File: frontend/lib/engines/sticks.ts
// Portage TypeScript du moteur des bâtonnets (backend/app/games/sticks.py), pour le jeu hors
// ligne (phase 3, palier 2). Mêmes règles, mêmes tests. Le serveur reste la source de vérité :
// une partie jouée hors ligne est rejouée et validée par lui à la synchronisation.

import type { SticksState } from "@/lib/api";

export const MIN_STICKS = 5;
export const MAX_STICKS = 50;
export const DEFAULT_STICKS = 21;
export const MAX_TAKE = 3;

export type Side = "player" | "ai";

export class IllegalMoveError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IllegalMoveError";
  }
}

export function other(side: Side): Side {
  return side === "player" ? "ai" : "player";
}

export function legalMoves(state: Pick<SticksState, "sticks" | "finished">): number[] {
  if (state.finished) return [];
  return Array.from({ length: Math.min(MAX_TAKE, state.sticks) }, (_, i) => i + 1);
}

export function newGame(sticks = DEFAULT_STICKS, first: Side = "player"): SticksState {
  if (!Number.isInteger(sticks) || sticks < MIN_STICKS || sticks > MAX_STICKS) {
    throw new RangeError(`sticks must be between ${MIN_STICKS} and ${MAX_STICKS}, got ${sticks}`);
  }
  const state = { sticks, current: first, winner: null, finished: false };
  return { ...state, legal_moves: legalMoves(state) };
}

export function apply(state: SticksState, take: number): SticksState {
  if (state.finished) throw new IllegalMoveError("game is finished");
  if (!legalMoves(state).includes(take)) {
    throw new IllegalMoveError(`cannot take ${take} sticks with ${state.sticks} remaining`);
  }
  const remaining = state.sticks - take;
  if (remaining === 0) {
    // Celui qui prend le dernier bâtonnet perd.
    return {
      sticks: 0,
      current: state.current,
      winner: other(state.current),
      finished: true,
      legal_moves: [],
    };
  }
  const next = { sticks: remaining, current: other(state.current), winner: null, finished: false };
  return { ...next, legal_moves: legalMoves(next) };
}

/** Vrai si celui qui doit jouer avec `sticks` bâtonnets perd face à un jeu parfait. */
export function isLosingPosition(sticks: number): boolean {
  return sticks % 4 === 1;
}

/** Coup qui laisse 4k+1 bâtonnets à l'adversaire, ou null en position perdante. */
export function winningMove(sticks: number): number | null {
  const take = (sticks - 1) % 4;
  return take >= 1 && take <= MAX_TAKE ? take : null;
}
